### DISCUSSION TITLE
esp32 special devices support issues

### DISCUSSION THREAD

**Discussion by gemu2015** [arendst/Tasmota#10263] [Q&A] [answered]
non generic esp32 devices like the m5stack core2 have a specific wiring to peripheral chips and the pin assignments of theses chips are also specific to this design. i don't see how to implement the drivers in a generic way. Also these devices sometimes use 2 different hardware spi, i2c and i2s busses.
to make this more generic we would need the GPIO definitions for 2 busses each and also more then one chip select for the spi busses.

but since we mostly need a specific driver for each device and it makes no sense to select different GPIOs as in generic devices we could also hardcode the driver with a specific device template, only leaving not used pins open for selecting GPIO functions.

these devices always need a self compile and they should be ready for use directly after compiling with all pins already assigned.

**Comment by blakadder** [comment, marked as answer]
I have a device with only some of the components used in m5stack and some of the components used in another device  but on different gpio's but I cannot use the available drivers correctly if they/re hardcoded and/or implemented all in one.

While true that almost every device will probably need a self compile with specific options this is what tasmota was at the beginning with modules but as the gamut of available devices grows larger you run into the same problem of trying to add every device into the code and maintain it at the same time

**Comment by gemu2015** [comment]
ok in a first step we would need 2 busses each and more than 1 SPI chip select. this will solve the problem partly but i don't see how to template very specific wiring as with the AXP power chips.

**Comment by arendst** [comment]
Regarding SPI chip select I noticed too we need specific CS for any SPI bus user. So every SPI display driver will need one in addition to MFRC522.

I will initiated this one of these days.

**Comment by gemu2015** [comment]
and also one for the SD card. 
there are also other SPI problems with special devices. some displays have only MOSI (no MISO) which means the driver will be disabled with the current implementation. i solved this by using SSPI and switching to hardware SPI if available. But Hardware spi without MISO should be possible too.
